### PR TITLE
[SPARK-39749][SQL] Always use plain string representation on casting Decimal to String

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

Currently, casting decimal as string type will result in Strings with exponential notations if the adjusted exponent is less than -6. This is consistent with BigDecimal.toString https://docs.oracle.com/javase/8/docs/api/java/math/BigDecimal.html#toString 

After this PR, the casting always uses plain string representation.

### Why are the changes needed?

1. The current behavior doesn't compliant to the ANSI SQL standard.
image
image

2. It is different from databases like PostgreSQL/Oracle/MS SQL server/etc.
3. The current behavior may surprise users since it only happens when the adjusted exponent is less than -6. The following query will return `false` by default (when ANSI SQL mode is off) since the `0.0000000123` is converted as `1.23E-7`:
```sql
select '0.000000123' in (0.000000123);
```

### Does this PR introduce _any_ user-facing change?

Yes, after changes, Spark SQL always uses plain string representation on casting Decimal to String. To restore the legacy behavior, which uses scientific notation if the adjusted exponent is less than -6, set `spark.sql.legacy.castDecimalToString.enabled` to `true`.

### How was this patch tested?

Unit test

### CODE REVIEW

**Comment by gengliangwang** [comment]
cc @timarmstrong @entong @cloud-fan @srielau

**Review by cloud-fan — APPROVED** [review]

**Review by HyukjinKwon — APPROVED** [review]

**Comment by gengliangwang** [comment]
Let's keep this open for one or two days since it is changing the behavior.

**Review by MaxGekk — APPROVED** [review]

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuiteBase.scala [hunk 1]
```diff
@@ -1305,4 +1305,12 @@ abstract class CastSuiteBase extends SparkFunSuite with ExpressionEvalHelper {
         Cast(child, DecimalType.USER_DEFAULT), it)
     }
   }
+
+  test("cast Decimal to string") {
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuiteBase.scala:1309]: nit. Could you add `SPARK-39749`?

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuiteBase.scala (hunk 1 above)
Comment (gengliangwang) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuiteBase.scala:1309]: Yes, sure

**Comment by gengliangwang** [comment]
@cloud-fan @HyukjinKwon @MaxGekk @dongjoon-hyun Thanks for the review. Merging to master.

**Pull request closed without merging**